src/kagglehub/packages.py

```diff
@@ -0,0 +1,125 @@
+    
+    def apply_to_module(self, module: ModuleType) -> ModuleType:
+        """Decorates all functions/methods in the module to apply our scope."""
+        def apply(obj: ModuleType | Type):
+            for name, member in inspect.getmembers(obj):
+                # Only decorate things defined within the module.
+                if getattr(member, "__module__", None) != module.__name__:
+                    continue
+                # These denylisted entries cause infinite loops
+                elif inspect.isclass(member) and name not in ["__base__", "__class__"]:
+                    apply(member)
+                elif inspect.isfunction(member) or inspect.ismethod(member):
+                    setattr(obj, name, decorate(member))
+        
+        def decorate(func: Callable) -> Callable:
+            @wraps(func)
+            def wrapper(*args, **kwargs):
+                with self:
+                    return func(*args, **kwargs)
+            return wrapper
+        
+        apply(module)
+
+
+logger = logging.getLogger(__name__)
+
+
+def package_import(handle: str, *, force_download: Optional[bool] = False) -> str:
+    """[WORK IN PROGRESS]
+
+    Download a Package (special type of Notebook) and import it.
+
+    Args:
+        handle: (string) the notebook handle under https://kaggle.com/code.
+        force_download: (bool) Optional flag to force download motebook output, even if it's cached.
+
+
+    Returns:
+        The imported python package.
+    """
+    h = parse_package_handle(handle)
+
+    logger.info(f"Downloading Notebook Output for Package: {h.to_url()} ...", extra={**EXTRA_CONSOLE_BLOCK})
+    package_path = registry.notebook_output_resolver(h, path=None, force_download=force_download)
+
+    # Unique module name based on handle
+    module_name = re.sub(r"[^a-zA-Z0-9_]", "_", f"kagglehub_package_{str(h)}")
+    init_file_path = pathlib.Path(package_path) / "package" / "__init__.py"
+    
+    spec = importlib.util.spec_from_file_location(module_name, init_file_path)
+    module = importlib.util.module_from_spec(spec)
+    sys.modules[module_name] = module
+    spec.loader.exec_module(module)
+    
+    return module
+
+
+def get_package_datasource_version_number(h: ResourceHandle) -> Optional[int]:
+    """Gets version number for given resource within current PackageScope.
+    
+    Returns None if no PackageScope is applied, or if it didn't contain the resource."""
+    scope = PackageScope.get()
+    return scope.datasources.get(h) if scope else None
+
+
+def get_package_asset_path(path):
+    """Returns a path referring to an asset file for use in Kaggle Packages.
+    
+    If within a PackageScope context, returns the path relative to it.
+    Otherwise, assumes we're in an interactive Kaggle Notebook where package
+    data should be written to the /kaggle/working dir."""
+    scope = PackageScope.get()
+    package_path = scope.path if scope else pathlib.Path("/kaggle/working/package")
```

Comment:
What if we are on colab or offline?

Reply:
Any time a Package is imported -- whether via `kagglehub.package_import` or directly, whether on Colab or offline -- the package's code should be applying its own `PackageScope` (through logic in the generated `__init__.py` file, with the helpers above in this file) so we assume it's only unset when you're defining the Package, currently intended only within a Kaggle Notebook, though we'll probably want to add more flexibility in the future.

Added a bit more docstring on this point, thanks.